### DISCUSSION TITLE
Task watchdog got triggered

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#10827] [ESP32] [answered]
Hi there,

I'm issuing a watchdog triggering but something is not so clear to me.

I'm using a Timer that feeds up the watchdog:

```
self.wdt = WDT(timeout=15000)
wdt_timer = Timer(1)
wdt_timer.init(mode=Timer.PERIODIC, period=4000, callback=self.__wd_timer_callback)
```

```
def __wd_timer_callback(self, timer):
    self.wdt.feed()
    print('[%d] WDT FEED \n' % time.ticks_ms())
```

The problem is that I always get this problem:

```
[33541] WDT FEED

E (37561) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (37561) task_wdt:  - mp_task (CPU 1)
E (37561) task_wdt: Tasks currently running:
E (37561) task_wdt: CPU 0: IDLE
E (37561) task_wdt: CPU 1: IDLE
E (37561) task_wdt: Aborting.

abort() was called at PC 0x4205dfe4 on core 0
```
The ticks associated to the problem is almost the same at every boot so I think the problem might be something in my code that I'm looking for.

What is not so clear to me is why I've got a WDT trigger at 37561 ms since I made a WDT feed at 33541 and the timeout is set to 15s.
If the feed was not working the correct way I expected a reset at ticks=15000 but I arrived to 37561

Hope you can help me. Thank you very much for your time

**Comment by davefes** [comment]
Does it happen consistently at 37561?  Are you going into light or deepsleep?

**Reply by ghost** [reply]
Yes, the time is almost the same. It may differs some ms. No, I'm not going into light or deepsleep.

**Reply by davefes** [reply]
At first I thought this looks like the software watchdog timer I use, but are you using machine.WDT?  

From https://docs.micropython.org/en/latest/library/machine.WDT.html

> Once started it cannot be stopped or reconfigured in any way.

I am not sure that this applies.

**Comment by karfas** [comment, marked as answer]
I can't reproduce this using a slightly modified version of your code on a ESP32-Wroom-32.
What else is running there ?
However, I assume the `print()` and/or the `time()` call _might_ allocate memory.
It might be more safe to use `micropython.schedule()`  from inside the timer callback.

Did you read the _fine_ documentation regarding `machine.Timer` on https://docs.micropython.org/en/latest/library/machine.Timer.html ?

By the way, I find it pretty useless to trigger a watchdog from within a hardware interrupt.

**Comment by ghost** [comment]
I've solved the problem NOT USING the Timer. It seems that something in the code is overriding the Timer. The "WDT FEED" should have confirmed the wdt feeding but it seems that the Timer callback didn't do that.
I've placed the feed into the main loop, apart from the Thread I'm creating into my code, and now it works.
Anyway it would be interesting in understanding which the cause it might be.

@davefes Yes, I'm using the `machine.WDT`

@karfas The whole code is quite huge, I have about 70 .py files, and it's quite hard to explain what else is running. In the main I'm creating 3 threads. It seems that the problem is created after the last thread is created, it handles an MQTT Client.
What is the problem in using `print()` or `time()`? Yes, they will allocate memory for sure but gc.collect() might clean it up. Anyway I'm using less then 1MB of RAM and the module has 8MB of (PS)RAM
What is your suggestion as an alternative to trigger the wdt from an an hardware interrupt?
Please can you explain to me better the `micropython.schedule(`) function?

Thank you all!

**Reply by karfas** [reply]
> What is your suggestion as an alternative to trigger the wdt from an an hardware interrupt?

Feeding the WDT from a timer interrupt *might* monitor that the timer and the processor is (in general) working and micropython is still (somewhat) alive, but **does not** prove that your application is really running and doing it's task.
The WDT feed belongs to the main loop of your application.

> Please can you explain to me better the micropython.schedule() function?

See documentation ( [Writing interrupt handlers](https://docs.micropython.org/en/latest/reference/isr_rules.html#isr-rules) ) and various discussions here and in the old forum.
`micropython.schedule()` is safe to be called from within an interrupt handler. 
The function given to `schedule()` will execute at the earliest possible time in context of the "normal" python runtime, not in interrupt context.

Regarding `print()`: **It's all in the documentation** I linked above.
From https://docs.micropython.org/en/latest/library/machine.Timer.html :
> See discussion of [important constraints](https://docs.micropython.org/en/latest/library/machine.html#machine-callbacks) on Timer callbacks.
Note
Memory can’t be allocated inside irq handlers (an interrupt) and so exceptions raised within a handler don’t give much information. See [micropython.alloc_emergency_exception_buf()](https://docs.micropython.org/en/latest/library/micropython.html#micropython.alloc_emergency_exception_buf) for how to get around this limitation.

The "important constraints" link gets me to
> A note of callbacks used by functions and class methods of [machine](https://docs.micropython.org/en/latest/library/machine.html#module-machine) module: all these callbacks should be considered as executing in an interrupt context
...
See [Writing interrupt handlers](https://docs.micropython.org/en/latest/reference/isr_rules.html#isr-rules).

In [Writing interrupt handlers](https://docs.micropython.org/en/latest/reference/isr_rules.html#isr-rules), I find 
> - Avoid memory allocation: no appending to lists or insertion into dictionaries, no floating point.
> - Consider using micropython.schedule to work around the above constraint.

**Reply by ghost** [reply]
Ok, really thank you for your support!
